fastapi/concurrency.py

```diff
@@ -1,40 +1,62 @@
 from contextlib import AsyncExitStack as AsyncExitStack  # noqa
 from contextlib import asynccontextmanager as asynccontextmanager
-from typing import AsyncGenerator, ContextManager, TypeVar
+from functools import partial
+from typing import Any, AsyncGenerator, ContextManager, Optional, TypeVar
 
 import anyio
-from anyio import CapacityLimiter
+from anyio.streams.memory import MemoryObjectReceiveStream, MemoryObjectSendStream
+from anyio.to_thread import run_sync
 from starlette.concurrency import iterate_in_threadpool as iterate_in_threadpool  # noqa
 from starlette.concurrency import run_in_threadpool as run_in_threadpool  # noqa
 from starlette.concurrency import (  # noqa
     run_until_first_complete as run_until_first_complete,
 )
 
-_T = TypeVar("_T")
+T = TypeVar("T")
+
+
+def _cm_thead_worker(
+    cm: ContextManager[T],
+    res_stream: MemoryObjectSendStream[T],
+    err_stream: MemoryObjectReceiveStream[Optional[Exception]],
+) -> None:
+    with cm as res:
+        anyio.from_thread.run(res_stream.send, res)
+        exc = anyio.from_thread.run(err_stream.receive)
+        if exc:
+            raise exc
+
+
+MaybeException = Optional[Exception]
 
 
 @asynccontextmanager
 async def contextmanager_in_threadpool(
-    cm: ContextManager[_T],
-) -> AsyncGenerator[_T, None]:
-    # blocking __exit__ from running waiting on a free thread
-    # can create race conditions/deadlocks if the context manager itself
-    # has it's own internal pool (e.g. a database connection pool)
-    # to avoid this we let __exit__ run without a capacity limit
-    # since we're creating a new limiter for each call, any non-zero limit
-    # works (1 is arbitrary)
-    exit_limiter = CapacityLimiter(1)
-    try:
-        yield await run_in_threadpool(cm.__enter__)
-    except Exception as e:
-        ok = bool(
-            await anyio.to_thread.run_sync(
-                cm.__exit__, type(e), e, None, limiter=exit_limiter
-            )
-        )
-        if not ok:
-            raise e
-    else:
-        await anyio.to_thread.run_sync(
-            cm.__exit__, None, None, None, limiter=exit_limiter
+    cm: ContextManager[T],
+    limiter: Optional[anyio.CapacityLimiter] = None,
```

Comment:
This is currently unused but IMO exposing it makes sense for future customizability. Happy to remove it if not